cli: handle utf in show tables

mrzasa opened the pull request:
Fixes: https://github.com/cockroachdb/cockroach/issues/51459 - UTF table names in `SHOW TABLES`

CLI on master shows UTF table names as escaped codepoints instead of UTF chars. It's because table name  is stored in a `relname` column an internal NAME type. Internally it's handled by a `DName` type that is sent to CLI as a byte array. This byte array is rendered as is, without UTF handling. See [sql_utl.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/sql_util.go#L1124-L1158)

This PR adds casting `relname` to `string` to achieve proper UTF handling in the CLI.

Alternatively we could convert `DName` to `string` before converting it to `driver.Value`. I'm not sure where to do it. If it's a better way to go, I'd be grateful for some guidance.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=56630) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=56630) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=56630) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56630)
<!-- Reviewable:end -->

knz commented:
Thanks for the contribution.
It's a bit surprising to see an ad-hoc solution for SHOW TABLES when many other SHOW statements return columns with (internal) type "name" to clients. Why should SHOW TABLES be an exception?

Moreover, the postgres protocol does have a native "name" type. So here's a strategy for you:

1. ensure that columns with internal type "name" also use the "name" postgres OID type when sent back to clients. Is that already the case? (I think so) If not, we can introduce this change -- I foresee it would be generally beneficial to all applications.


2. in the `cockroach sql` code (client-side), inspect the column type of returned results. Is the data still received as `[]byte` even when the server-side type is "name"? If yes, then it's possible to be smarter about the representation. This is possible because the type of columns is available via `rows.ColumnTypes()`. You could add code to inspect this column type and decide to display the value differently if the type is "name".

mrzasa commented:
@knz thanks for your reply. I agree that SQL hacking is not the best solution. With your hints I should be able to prepare something more general.

mrzasa committed:
cli: render NAME results without octal escapes

Previously internal Name type with non-ASCII characters wwas shown
incorrectly in CLI (e.g. SHOW TABLES, SHOW CONSTRAINTS).
For instance: '\305\274\303\263\305\202w' instead of 'żółw'.
It caused inconvenience while using CLI.

To fix it this patch exposes ColumnTypeDatabaseTypeName in sqlRowsI
interface then uses this function to determine whether returned data is
of a Name type. If it's the case it's casted from []bytes to a string.

Release note (bug fix): Non-ASCII characters in NAME results in
`cockroach sql`  / `demo`,
(e.g. in the results SHOW TABLES, SHOW CONSTRAINTS) are now displayed
without being escaped to octal codes.

mrzasa commented:
@knz I've updated my patch:
* I've identified another case when the name is displayed incorrecty - `SHOW CONSTRAINTS`, I've added it to the test case
* I've used `ColumnTypeDatabaseTypeName` that returns a textual name of a database type. To be able to use it, I've extended `sqlRowsI` with `driver.RowsColumnTypeDatabaseTypeName`. 
* I tried adding  `ColumnTypes()` to `sqlRowsI`, but it eneded up with  a `anic: interface conversion: *pq.rows is not cli.sqlRowsI: missing method ColumnTypes`. I believe the reason for that is `pq` not implementing this function: https://github.com/lib/pq/blob/master/rows.go
* I've used the type name returned from `ColumnTypeDatabaseTypeName` and it's `NAME` I cast the data to string.

I'll be grateful for feedback.

mrzasa commented:
Folks, do you need anything more form me at this stage? I'll be happy to improve the solution if it fails to meed the standards.

knz reviewed (approved):
This is a very good contribution. Thank you!

bors r+

> Reviewed 2 of 2 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56630#-:-MRK9bl-3-i7rRhZjFUT:bxe9ffk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

knz commented:
I'll have a look.

mrzasa commented:
New column `locality` was added to `SHOW TABLES` output. I think it's safe to add the new column to the expected output.

knz commented:
will do, thanks

knz commented:
bors r+
